056-zk-kraft-migration.md

```diff
@@ -0,0 +1,435 @@
+
+**_Preconditions_**
+
+The FSM is in the `Migration` state (as per `Kafka.status.migrationState` field).
+The migration is ongoing.
+
+**_Trigger_**
+
+A periodic reconciliation is triggered.
+
+**_Actions_**
+
+The operator checks the Kafka migration metrics by using the Kafka Agent client in order to verify the status of the migration on the Apache Kafka cluster.
+The metric to be used for the check is the `ZkMigrationState` which stays in `PRE_MIGRATION` while the actual migration is still going on and moves to `MIGRATION` when it's done. More details in the "Metrics" section.
+The current reconciliation ends.
+
+**_Status changes_**
+
+If the migration finished, the FSM moves to the `DualWrite` state.
+If the migration is still ongoing, the FSM stays in the `Migration` state.
+
+---
+
+While migration is running, it's still possible to not move forward with finalizing the migration but rolling back instead.
+
+**_Preconditions_**
+
+The FSM is in the `Migration` state (as per `Kafka.status.migrationState` field).
+The migration is ongoing.
+
+**_Trigger_**
+
+The user can rollback the process by applying the `strimzi.io/kraft: disabled` annotation on the `Kafka` custom resource.
+
+**_Actions_**
+
+In this case, the operator starts a reconciliation an, via the `KafkaReconciler` component, configures the brokers with no ZooKeeper migration flag, and no KRaft controllers quorum connection information.
+Finally, the operator rolls the brokers, via the `KafkaRoller`, and waits for them to be ready again.
+The current reconciliation ends.
+
+**_Status changes_**
+
+The FSM moves back to the `ControllersInMigration`, so actually at the beginning of the Phase 2.
+
+---
+
+When the controllers are configured with ZooKeeper migration enabled but the user has applied the `strimzi.io/kraft: disabled` annotation on the `Kafka` custom resource during the migration, the brokers are working with the ZooKeeper ensemble again and the user can rollback the overall process by deleting the controllers as well.
+
+**_Preconditions_**
+
+The `strimzi.io/kraft: disabled` annotation is already applied on the `Kafka` custom resource.
+The FSM is in the `ControllersInMigration` state (as per `Kafka.status.migrationState` field) and keep staying in this state until the user deletes the controllers pool(s) to rollback (or re-apply `strimzi.io/kraft: migration` to move forward with migration).
+
+**_Trigger_**
+
+The user deletes the `KafkaNodePool` custom resource(s) defining the controllers already running with ZooKeeper migration enabled.
+
+**_Actions_**
+
+In this case, the operator starts a reconciliation an deletes the pods running the controllers.
+The current reconciliation ends.
+
+**_Status changes_**
+
+The FSM moves back to the `ZooKeeper`, so actually at the beginning of the Phase 1.
+There is no more migration in progress and the Apache Kafka cluster is running in ZooKeeper mode.
+
+#### Phase 3/4: Migrating brokers and controllers to full KRaft
```

Comment:
In the titles you have `Phase 1/2` and `Phase 3/4` => I would read it as _phase 1 out of 2_ and _phase 3 out of 4_. But that does not make sense. Should we rename it? Either just call it phase 1 and 2, or `Phase 1 and 2` and `Phase 3 and 4`?

Reply:
Yeah, agree.